### ISSUE TITLE
source code for the Principled shader 

### ISSUE THREAD

**Issue by BluePrintRandom** [youle31/Ketsji-help#2]
https://github.com/dfelinto/blender/blob/master/source/blender/gpu/shaders/gpu_shader_material.glsl#L2628

so this node gives us PBR in the viewport,  and I think is already in our sources.

is there any way you could help us understand why the material is not present in the GE?

**Comment by youle31** [comment]
Hi, this is for cycles only in 2.7, no? Cycles doesn't use the same shaders than bge.
eevee implements cycles shaders for the viewport (realtime). So interactive mode will have principled bsdf.

**Comment by BluePrintRandom** [comment]
The shader works in the viewport as it is now to a limited extent

do cycles, add principled node
swap to game engine add output node - connect output from principled node
to output

metalness etc work
the scenee light is not updated currently 

we need lightfields as well for spec/ reflect / IBL

**Comment by youle31** [comment]
Ok... weird. I have no idea about how it works... if i have motivation, i'll try to reproduce what you say and have a look at the code.

**Comment by youle31** [comment]
@BluePrintRandom : the principled bsdf looks like that: http://pasteall.org/pic/show.php?id=78d9b02f18e8d5030c2f5b12c07c5e0d with default settings on the default cube. You mean that you can have this result elsewhere than in cycles?

**Comment by BluePrintRandom** [comment]
yeah this is all the way back in 2.1/ge normals
https://youtu.be/h7R0QWJsZgo

many of the cycles texture nodes work in the bge as well
(brick texture etc)

**Comment by youle31** [comment]
Okok... i don't really know... seems it is not influenced by light position

**Comment by BluePrintRandom** [comment]
yeah they are set in preferances*
we need to hijack this code and overide it*

**Comment by youle31** [comment]
It seems it is just the cycles material preview in viewport. I don't know how it works...

**Comment by youle31** [comment]
if you want to access cycles materials preview in game engine viewport, you can look at: external_engines.c

"BLENDER_GAME", N_("Blender Game"), RE_INTERNAL | RE_GAME,

and change it with:

"BLENDER_GAME", N_("Blender Game"), RE_INTERNAL | RE_GAME | RE_USE_SHADING_NODES,

But it's just material previsualisation (doesn't take into account lighting)

**Comment by BluePrintRandom** [comment]
yeah but this location is where we can define and use PBr in the ge
(we overide this node or copy data from it and feed a proper PBR pipeline
without mucking around a bunch of ui)
